File: static/js/notification.js
document.addEventListener("DOMContentLoaded", function () {
  const notificationsToggle = document.querySelector(".notifications-toggle");
  const notificationsPanel = document.querySelector(".notifications-panel");
  const notificationsList = document.getElementById("notifications-list");
  const notificationBadge = document.getElementById("notification-badge");
  const markAllReadBtn = document.querySelector(".mark-all-read");

  // Toggle notifications panel
  notificationsToggle.addEventListener("click", function (e) {
    e.stopPropagation();
    notificationsPanel.classList.toggle("show");
  });

  // Load notifications
  function loadNotifications() {
    fetch("/notifications")
      .then((response) => response.json())
      .then((notifications) => {
        if (notifications.length === 0) {
          notificationsList.innerHTML = `
            <div class="empty-state">
              <i class="fas fa-bell-slash"></i>
              <p>No new notifications</p>
            </div>
          `;
          return;
        }

        notificationsList.innerHTML = notifications
          .map(
            (notification) => `
          <div class="notification ${notification.is_read ? "" : "unread"}" 
               data-id="${notification.id}" 
               data-type="${notification.type}">
            <div class="notification-icon">
              ${getNotificationIcon(notification.type)}
            </div>
            <div class="notification-content">
              <h5>${notification.title}</h5>
              <p>${notification.message}</p>
              <small>${formatTime(notification.created_at)}</small>
            </div>
          </div>
        `
          )
          .join("");

        updateBadgeCount();

        // Add click handlers
        document.querySelectorAll(".notification").forEach((notification) => {
          notification.addEventListener("click", function () {
            if (this.classList.contains("unread")) {
              markNotificationRead(this.dataset.id);
              this.classList.remove("unread");
              updateBadgeCount();
            }
            // Handle navigation based on type
            handleNotificationClick(this.dataset.type, this.dataset.id);
          });
        });
      });
  }

  // Mark all as read
  markAllReadBtn.addEventListener("click", function (e) {
    e.stopPropagation();
    fetch("/notifications/read-all", { method: "POST" })
      .then((response) => response.json())
      .then((data) => {
        if (data.success) {
          document.querySelectorAll(".notification.unread").forEach((el) => {
            el.classList.remove("unread");
          });
          updateBadgeCount();
        }
      });
  });

  // Helper functions
  function getNotificationIcon(type) {
    const icons = {
      prescription: '<i class="fas fa-prescription-bottle-alt"></i>',
      diagnosis: '<i class="fas fa-stethoscope"></i>',
      payment: '<i class="fas fa-money-bill-wave"></i>',
      inventory: '<i class="fas fa-pills"></i>',
      appointment: '<i class="fas fa-calendar-check"></i>',
      patient: '<i class="fas fa-user-injured"></i>',
      staff: '<i class="fas fa-user"></i>',
      medicine: '<i class="fas fa-capsules"></i>',
      disease: '<i class="fas fa-disease"></i>',
    };
    return icons[type] || icons["system"];
  }

  function formatTime(isoString) {
    const date = new Date(isoString);
    return date.toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" });
  }

  function updateBadgeCount() {
    const unreadCount = document.querySelectorAll(
      ".notification.unread"
    ).length;
    notificationBadge.textContent = unreadCount;
    notificationBadge.style.display = unreadCount > 0 ? "flex" : "none";
  }

  function markNotificationRead(id) {
    fetch(`/notifications/${id}/read`, { method: "POST" });
  }

  setInterval(loadNotifications, 1000);

  updateBadgeCount();
});

File: Models/notification.py
from Models.base_model import db, BaseModel, get_local_time
from enum import Enum

class NotificationType(Enum):
  PRESCRIPTION = 'prescription'
  DIAGNOSIS = 'diagnosis'
  PAYMENT = 'payment'
  INVENTORY = 'inventory'
  APPOINTMENT = 'appointment'
  PATIENT = 'patient'
  STAFF = 'staff'
  MEDICINE = 'medicine'
  DISEASE = 'disease'
  # LABTEST = 'labtest'

class Notification(BaseModel, db.Model):
  clinic_id = db.Column(db.Integer, db.ForeignKey('clinic.id'), nullable=False)
  notification_type = db.Column(db.Enum(NotificationType), nullable=False)
  title = db.Column(db.String(100), nullable=False)
  message = db.Column(db.String(255), nullable=False)
  is_read = db.Column(db.Boolean, default=False)
  related_id = db.Column(db.Integer)
  created_at = db.Column(db.DateTime, default=get_local_time())

  def to_dict(self):
    return {
      'id': self.id,
      'type': self.notification_type.value,
      'title': self.title,
      'message': self.message,
      'is_read': self.is_read,
      'created_at': self.created_at.isoformat(),
      'related_id': self.related_id
    }
